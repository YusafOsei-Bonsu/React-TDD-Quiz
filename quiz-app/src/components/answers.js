import React from 'react'

const Answers = (props) => {
    // let answers = [props.results.correct_answer];
    let answers = 'answers'
    // answers = answers.concat(props.results.incorrect_answers);
    // answers.sort(function(a, b){return 0.5 - Math.random()});
    
    return ( 
        <div className='answers'>
            <p>{answers[0]}</p>
            <p>{answers[1]}</p>
            <p>{answers[2]}</p>
            <p>{answers[3]}</p>
        </div>
    );
}

export default Answers;