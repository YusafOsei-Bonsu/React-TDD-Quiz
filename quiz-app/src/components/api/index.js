import axios from 'axios';
export const getCatData = (dispatch) => {
    return (dispatch) => {
        axios.get("https://opentdb.com/api_category.php").then(response => {
            const data = response.data;
            dispatch({ type: 'catDropDown', payload: data.trivia_categories });
        })

    }
};
// Fetch for quiz by category id, level and 
export const getQuizData = (value, dif, dispatch) => {
    return (dispatch) => {
        axios.get(`https://opentdb.com/api.php?amount=10&category=${value}&difficulty=${dif}&type=multiple`).then(async response => {
            const data = response.data.results;
            await dispatch({ type: 'questionsAnswers', payload: data });
        }).catch(() => dispatch({type: 'handle-error' }))
    }
};