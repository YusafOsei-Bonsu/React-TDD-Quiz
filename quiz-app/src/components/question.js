import React from 'react'

const Question = () => {
    return (
        <div className='question'>
           <p>Question</p> 
        </div>
    );
}

export default Question;