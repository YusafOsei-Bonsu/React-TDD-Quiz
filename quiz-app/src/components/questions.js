import React from 'react'
const Questions = ({results}) => {
  console.log(results)
    return (
        <div className='answers'> 
          <h5>{atob(results.question)}</h5>
        </div>
        );
}
export default Questions;