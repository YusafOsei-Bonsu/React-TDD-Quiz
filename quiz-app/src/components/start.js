import React from 'react';
import '../styles/start.css';

const Start = () => {
    return (
        <div className='start'>
           <a href="/form"><button className="glow-on-hover">Start Quiz</button></a> 
        </div>
    )
}

export default Start;