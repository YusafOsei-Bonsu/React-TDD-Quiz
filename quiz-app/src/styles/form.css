.form {
    display: flex;
    flex-direction: column;
    background-color: #F1F0EA;
    width: 20%;
    margin-top: 250px;
    margin-left: auto;
    margin-right: auto;
    padding: 10px;
    justify-content: left;
}

input[type="text"] {
    padding: 5px;
}

input[type="submit"][value="Go"] {
    cursor: pointer;
    font-size: 25px;
    position: relative;
    z-index: 0;
    border: none;
    outline: none;
    background-color: rgb(0, 0, 0);
    color: rgb(255, 255, 255);
    border-radius: 10px;
    padding-top: 5px;
    padding-bottom: 5px;
}

input[type="submit"][value="Go"]:hover {
    color: rgb(0, 0, 0);
    background-color: rgb(255, 255, 255);
    border: 2px solid rgb(0, 0, 0);
}

input[type="submit"][value="Go"]:active {
    color: rgb(255, 255, 255);
    background-color: rgb(169, 169, 169);
    border: 2px solid rgb(255, 255, 255);
}