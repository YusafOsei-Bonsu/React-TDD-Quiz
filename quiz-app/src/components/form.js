import React from 'react';
import '../styles/form.css';
import { connect } from 'react-redux';
import { getQuizData } from './api/';
import { useHistory } from 'react-router-dom';

export const Form = (props) => {
    let info = props.categories;
    // remove link that never loads
    info.splice(20, 1)
    let data = props.quizData
    let history = useHistory();
    return (
        <div>
            <form className='form' onSubmit={(event) => { props.handleSubmit(event, history, data) }}>
                <select name="topic">
                    {info.map((data) => { return <option className={data.id} key={data.id} value={data.id}>{data.name}</option> })}
                </select>
                <br />
                <label htmlFor="easy">Easy</label>
                <input id="easy" type="radio" key="easy" name="difficulty" value="easy" />
                <br />
                <label htmlFor="medium"><input id="medium" type="radio" name="difficulty" value="medium" />Medium</label>
                <br />
                <label htmlFor="hard"><input id="hard" type="radio" name="difficulty" value="hard" />Hard</label>
                <br />
                <input id="textInput" onChange={props.handleChange} type='text' placeholder='player1' /><br />
                <input type="submit" value="Go" />
            </form>
        </div>
    )
}
export const mapStateToProps = (state, ownProps) => {
    return { categories: state.catDropDown, justStore: ownProps, userName: state.users, quizData: state.quizData }
};
//add an on change to add user name to data
export const mapDispatchToProps = (dispatch) => {
    return {
        handleSubmit: (event, history, data) => {
            event.preventDefault();
            let difficulty = event.target.difficulty.value;
            let topic = event.target.topic.value;
            dispatch({type: 'reset'});
            dispatch(getQuizData(topic, difficulty));
            if (data.length === 0) {
                alert('Server Error. Please try again')
                history.push(`/form/`)
            }
            else {
                history.push('/quiz/0');
            }
        },
        handleChange: (event) => {
            let name = event.target.value;
            dispatch({ type: 'addUser', payload: name });
        }
    }
}
export default connect(mapStateToProps, mapDispatchToProps)(Form);