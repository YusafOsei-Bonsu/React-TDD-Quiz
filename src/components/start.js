import React from 'react';
import { useHistory } from 'react-router-dom';
import '../styles/start.css';

const Start = () => {
    let history = useHistory();

    // Navigates to the form page
    function goToForm() {
        history.push('/form');
    }

    return (
        <div className='start'>
           <button className="glow-on-hover" onClick={goToForm}>Start Quiz</button>
        </div>
    )
}

export default Start;