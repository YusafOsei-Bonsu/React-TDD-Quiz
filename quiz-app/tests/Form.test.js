import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import { Form , mapStateToProps} from '../src/components/form';
import configureMockStore from 'redux-mock-store';
import { connect } from 'react-redux';
import mockAxios from 'axios';
import thunk from 'redux-thunk';
import { createPromise, promiseMiddleware } from 'redux-promise-middleware';
import { isMainThread } from 'worker_threads';
import { useHistory } from 'react-router-dom'

jest.mock('react-router-dom', () => ({
    useHistory: () => ({
        push: jest.fn(),
    }),
}));

function setup() {
    const props = {
        categories: [
            { bad: 'good' },
            { "id": 9, "name": "General Knowledge" },
            { "id": 10, "name": "Entertainment: Books" }
        ],
        handleChange: jest.fn(),
        handleSubmit: jest.fn()

    }

    const enzymeWrapper = shallow(<Form {...props} />)
    const enzymeWrapperM = mount(<Form value='target' {...props} />)

    return {
        props,
        enzymeWrapper,
        enzymeWrapperM
    }
}

describe('Form Component', () => {

    it('should render self and subcomponents', () => {
        const { enzymeWrapper } = setup()

        expect(enzymeWrapper.find('form').hasClass('form')).toBe(true)

    })
    it('calls handleSubmit prop function when form is submitted', () => {
        const { enzymeWrapperM, props } = setup()
        const form = enzymeWrapperM.find('form');
        form.simulate('submit');
        expect(props.handleSubmit).toHaveBeenCalledTimes(1);
    });
    it('calls handleChange', () => {
        const { enzymeWrapperM, props } = setup()
        const input = enzymeWrapperM.find('#textInput')
        input.simulate('change', {target: {value: 'church'}})
        input.simulate('keydown', {key: 'ArrowDown'})
        input.simulate('keydown', {key: 'Enter'})
        expect(props.handleChange).toHaveBeenCalledTimes(1)
    })
    it('should call onChange', () => {
        // const handleChange = jest.fn()
        const event = {
            preventDefault() {},
            target: {value: 'the-value'}
        };

        const { enzymeWrapper, props} = setup()
        const component = enzymeWrapper
        component.find('#textInput').simulate('change', event);
        expect(props.handleChange).toBeCalledWith(event);

    });

    it

    // // Checks if the drop-down box renders
    // it('Dropdown box should render', () => {
    //     const {enzymeWrapper} = setup()

    //     expect(enzymeWrapper.find('select').length).toEqual(1);
    // })
    // it('should call onChange if input text is greater than 0', () => {
    //     const {enzymeWrapper, props} = setup()
    //     const input = enzymeWrapper.find('hohoho')
    //     input.props().onSave('')
    //     expect(props.handleChange.mock.calls.length).toBe(0)
    // })



});

