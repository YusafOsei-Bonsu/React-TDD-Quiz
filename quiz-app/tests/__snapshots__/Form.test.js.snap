// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component MyComponent renders with default props 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <select
      name="topic"
    >
      <option />
      <option
        className={9}
        key="9"
        value={9}
      >
        General Knowledge
      </option>
      <option
        className={10}
        key="10"
        value={10}
      >
        Entertainment: Books
      </option>
    </select>
    <br />
    <label
      htmlFor="easy"
    >
      Easy
    </label>
    <input
      id="easy"
      name="difficulty"
      type="radio"
      value="easy"
    />
    <br />
    <label
      htmlFor="medium"
    >
      Medium
    </label>
    <input
      id="medium"
      name="difficulty"
      type="radio"
      value="medium"
    />
    <br />
    <label
      htmlFor="hard"
    >
      Hard
    </label>
    <input
      id="hard"
      name="difficulty"
      type="radio"
      value="hard"
    />
    <br />
    <input
      id="textInput"
      onChange={[MockFunction]}
      placeholder="player1"
      type="text"
    />
    <br />
    <input
      type="submit"
      value="Go"
    />
  </form>
</div>
`;
